### PR TITLE
Adding new feature wire_options to draw_mpl function 

### PR DESCRIPTION
**Pull request by zazabap**
**Context:**
On the level of mid-scale algorithms, it might be nice to differentiate between different "types" of wires, for example by coloring them differently, or giving them distinct line styles. As an example, in https://github.com/PennyLaneAI/qml/pull/1185 on preparing matrix product states, there are auxiliary bond qubits and physical qubits, and coloring them differently would be a neat thing to do.

**Description of the Change:**
Update the output wire_options that could change the line style and color for circuit output. 
```python
        @qml.qnode(qml.device("default.qubit"))
        def node(x):
            for w in range(5):
                qml.Hadamard(w) 
            return qml.expval(qml.PauliZ(0) @ qml.PauliY(1))

        # Make all wires cyan and bold, 
        # except for wires 2 and 6, which are dashed and another color
        wire_options = {"color": "cyan", 
                        "linewidth": 5, 
                        2: {"linestyle": "--", "color": "red"}, 
                        6: {"linestyle": "--", "color": "orange"}
                    }
        _,ax  = qml.draw_mpl(node, wire_options=wire_options)(0.52)
```

**Benefits:**
When complicated sates and quantum circuits diagram are created, wires could be marked with different selections.  

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/6165

### CODE REVIEW

**Comment by zazabap** [comment]
Dear @dwierichs, I made necessary changes here for the test cases, documentation, and source code. However, I am no so familiar with the format check here. Sorry maybe I need some additional help on it.

**Comment by dwierichs** [comment]
Hi @zazabap,
Great, thank you for your contribution. Is it ready to be reviewed? :) 
Regarding the continuous integration checks, including formatting checks, take a look at our [guide on how to submit a PR](https://docs.pennylane.ai/en/stable/development/guide/pullrequests.html).
In particular, you want to `pip install -r requirements-dev.txt` while in the PennyLane directory and then run
`black -l 100 pennylane tests`. Other formatting tests likely will not be super relevant in this PR (unless pylint complains, but this we can handle once we get there :) )
Hope this helps! Let me know when we can move to review!

**Comment by zazabap** [comment]
Thanks a lot for the help! I have reformatted the modified files and commit again, I think not it is ready for merge.

File: pennylane/drawer/mpldrawer.py [hunk 1]
```diff
@@ -293,12 +293,36 @@ def __init__(self, n_layers, n_wires, c_wires=0, wire_options=None, figsize=None
...
         if wire_options is None:
             wire_options = {}
+            self._wire_lines = [
+                plt.Line2D((-1, self.n_layers), (wire, wire), zorder=1, **wire_options)
+                for wire in range(self.n_wires)
+            ]
+        else:
+            # Separate default options from specific wire options
+            default_options = {k: v for k, v in wire_options.items() if isinstance(k, str)}
+            wire_specific_options = {k: v for k, v in wire_options.items() if isinstance(k, int)}
```
Comment (dwierichs) [inline pennylane/drawer/mpldrawer.py:303]: I should have mentioned this earlier on in the examples I think.
Wire labels are allowed to be any hashable object, so in particular strings are allowed.
I would therefore suggest to make the distinction of whether something is a per-wire option based on whether the key is in the wire labels. 
Now, this class does not even receive the wires but only the number of wires via `n_wires`, which would require us to change the interface of the class. I'd like to avoid this.
So as a surrogate, we could check whether the _value_ is a dictionary, and infer that if it is one, the key-value pair must've been a per-wire option. I'm not sure whether there are any line properties that themselves would have dictionaries as values, but it seems like the lesser evil at this point.

I think we should refine this somewhat hacky way of doing things later on, but for now I'd say this is a fair tradeoff between feature addition and tiny edge case (in which a dictionary-valued kwarg would be interpreted as per-wire option although it is meant to be an all-wires option).

Comment (dwierichs) [reply pennylane/drawer/mpldrawer.py:303]: Sorry for not seeing this one coming earlier!

File: tests/drawer/test_draw_mpl.py [hunk 2]
```diff
@@ -328,6 +328,49 @@ def test_wire_options(self):
...
+
+        for w in ax.lines:
+            assert w.get_color() == "orange"
+
+        # Wires are orange and cyan
+        wire_options = {0: {"color": "orange"}, 1: {"color": "cyan"}}
+        _, ax = qml.draw_mpl(f_circ, wire_options=wire_options)(0.52)
+
+        assert ax.lines[0].get_color() == "orange"
+        assert ax.lines[1].get_color() == "cyan"
```
Comment (dwierichs) [inline tests/drawer/test_draw_mpl.py:351]:
```suggestion
        assert ax.lines[1].get_color() == "cyan"
        assert ax.lines[2].get_color() == "black"
```

File: tests/drawer/test_draw_mpl.py [hunk 3]
```diff
@@ -328,6 +328,49 @@ def test_wire_options(self):
...
+        for i, w in enumerate(ax.lines):
+            if i == 2:
+                assert w.get_color() == "red"
+                assert w.get_linestyle() == "--"
+            elif i == 6:
+                assert w.get_color() == "orange"
+                assert w.get_linestyle() == "--"
+            else:
+                assert w.get_color() == "cyan"
+                assert w.get_linewidth() == 5
```
Comment (dwierichs) [inline tests/drawer/test_draw_mpl.py:372]: Just to make sure there are no leaked kwargs, optional suggestion:

```suggestion
                assert w.get_linestyle() == "--"
                assert w.get_linewidth() == 1
            elif i == 6:
                assert w.get_color() == "orange"
                assert w.get_linestyle() == "--"
                assert w.get_linewidth() == 1
            else:
                assert w.get_color() == "cyan"
                assert w.get_linestyle() == "-"
                assert w.get_linewidth() == 5
```

**Review by dwierichs — CHANGES_REQUESTED** [review]
Amazing addition @zazabap !
Looking forward to getting this feature in! :100:
I had some comments, in particular the documentation update should be moved to the docstrings of the respective functions, and the criterion with which we determine whether something is a per-wire option needs to be modified.

Thanks again, and let me know any questions you might have!

File: tests/drawer/test_draw_mpl.py (hunk 3 above)
Comment (zazabap) [reply tests/drawer/test_draw_mpl.py]: In this test case, universal line width is 5 and color is cyan. Only specific line 2 and line 6 have linewidth of 1. Therefore, I slightly modify the wire_options for the new testing case.

File: pennylane/drawer/mpldrawer.py [hunk 4]
```diff
@@ -293,12 +293,36 @@ def __init__(self, n_layers, n_wires, c_wires=0, wire_options=None, figsize=None
...
+
+            # Define defaults if "color" and "linewidth" aren't specified in default_options
+            default_color = default_options.get("color", "black")
+            default_linewidth = default_options.get("linewidth", 1)
```
[the remark this answers is not in the document]
Comment (zazabap) [reply pennylane/drawer/mpldrawer.py:307]: Hi, this is not used as the default case, it is used for some cases are not specified by the users. In some case like this:
`
@qml.qnode(qml.device("default.qubit"))
def node(x):
    for w in range(10):
         qml.Hadamard(w) 
    return qml.expval(Z(0) @ Y(1))

wire_options = {"linewidth": 5, 2: {"linestyle": "--", "color": "red"}, 2: {"linestyle": "--", "color": "orange"}}
qml.draw_mpl(node, wire_options=wire_options)(0.52)
`

for example if there is no input from the user to setup output's wire color, then the default is "black", I guess there should be a better naming for this part of the code so I rename it as global_color & global_linewidth.

File: pennylane/drawer/mpldrawer.py [hunk 5]
```diff
@@ -293,12 +293,36 @@ def __init__(self, n_layers, n_wires, c_wires=0, wire_options=None, figsize=None
...
+                line_options = {**default_options, **specific_options}
+
+                # Create Line2D with the combined options
+                line = plt.Line2D(
+                    (-1, self.n_layers),
+                    (wire, wire),
+                    zorder=1,
+                    color=line_options.get("color", default_color),
+                    linewidth=line_options.get("linewidth", default_linewidth),
+                    linestyle=line_options.get("linestyle", "-"),
```
[the remark this answers is not in the document]
Comment (zazabap) [reply pennylane/drawer/mpldrawer.py:322]: Update with previous comment

File: pennylane/drawer/mpldrawer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/drawer/mpldrawer.py:307]: I think I understand what you mean. However, if we simply do not pass any `color`, for example, Matplotlib will just draw a black line anyways, right?

File: doc/code/qml_drawer.rst [hunk 6]
```diff
@@ -50,6 +50,54 @@ deleting the ``fontlist`` file in the Matplotlib cache directory) — to find th
 cache directory, run ``matplotlib.get_cachedir()``. If this font is not
 available, the drawer will fall back on a default font.
 
+Wire Options
```
[the remark this answers is not in the document]
Comment (zazabap) [reply doc/code/qml_drawer.rst:53]: Thanks a lot! If possible, could you help me move this to the docstring? I am not so familiar with whether the specific documentation located.

File: pennylane/drawer/mpldrawer.py (hunk 1 above)
Comment (zazabap) [reply pennylane/drawer/mpldrawer.py]: Thanks a lot for the comment, I have modified the necessary part into dictionary checks.

File: doc/code/qml_drawer.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply doc/code/qml_drawer.rst]: I moved things to the docstrings of `draw_mpl` and `tape_mpl` and added the required figures to the static files.

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -6,7 +6,9 @@
 
 <h3>Improvements 🛠</h3>
 
-<h4>Other Improvements</h4>
+* Added support for the `wire_options` dictionary to customize wire line formatting in `qml.draw_mpl` circuit
+  visualizations, allowing global and per-wire customization with options like `color`, `linestyle`, and `linewidth`.
+[(#6486)](https://github.com/PennyLaneAI/pennylane/pull/6486)
```
Comment (dwierichs) [inline doc/releases/changelog-dev.md:11]:
```suggestion
  [(#6486)](https://github.com/PennyLaneAI/pennylane/pull/6486)
```

File: pennylane/drawer/mpldrawer.py [hunk 8]
```diff
@@ -293,12 +293,32 @@ def __init__(self, n_layers, n_wires, c_wires=0, wire_options=None, figsize=None
...
+                line_options = {**global_options, **specific_options}
+
+                # Create Line2D with the combined options
+                line = plt.Line2D(
+                    (-1, self.n_layers),
+                    (wire, wire),
+                    zorder=1,
+                    color=line_options.get("color"),
+                    linewidth=line_options.get("linewidth"),
+                    linestyle=line_options.get("linestyle"),
```
Comment (dwierichs) [inline pennylane/drawer/mpldrawer.py:318]:
```suggestion
                    **line_options,
```

File: pennylane/drawer/tape_mpl.py [hunk 9]
```diff
@@ -328,7 +330,7 @@ def tape_mpl(
         measurements = [qml.expval(qml.Z(0))]
         tape = qml.tape.QuantumTape(ops, measurements)
 
-        fig, ax = tape_mpl(tape)
+        fig, ax = qml.drawer.tape_mpl(tape)
```
Comment (dwierichs) [inline pennylane/drawer/tape_mpl.py:333]: Added the module reference to the function throughout the docstring to make it easier accessible when copying code from docstring.

File: tests/drawer/test_draw_mpl.py [hunk 10]
```diff
@@ -328,6 +328,75 @@ def test_wire_options(self):
...
+                assert w.get_linestyle() == "--"
+                assert w.get_linewidth() == 1
+            elif i == 6:
+                assert w.get_color() == "orange"
+                assert w.get_linestyle() == "--"
+                assert w.get_linewidth() == 1
+            else:
+                assert w.get_color() == "cyan"
+                assert w.get_linestyle() == "-"
+                assert w.get_linewidth() == 5
```
Comment (dwierichs) [inline tests/drawer/test_draw_mpl.py:376]: Sorry for being confused about the interaction of the different settings here! 
After actually understanding what is going on, I think the optimal test case would be like this, once inheriting the linewidth from global settings to wire 2 and once overwriting the global settings for wire 6.

Sorry for the back and forth! 

```suggestion
            2: {"linestyle": "--", "color": "red"},
            6: {"linestyle": "--", "color": "orange", "linewidth": 1},
        }
        _, ax = qml.draw_mpl(f_circ, wire_options=wire_options)(0.52)

        for i, w in enumerate(ax.lines):
            if i == 2:
                assert w.get_color() == "red"
                assert w.get_linestyle() == "--"
                assert w.get_linewidth() == 5
            elif i == 6:
                assert w.get_color() == "orange"
                assert w.get_linestyle() == "--"
                assert w.get_linewidth() == 1
            else:
                assert w.get_color() == "cyan"
                assert w.get_linestyle() == "-"
                assert w.get_linewidth() == 5
```

**Review by dwierichs — COMMENTED** [review]
Had some small follow up comments. Sorry for being confused :D

**Comment by zazabap** [comment]
Dear @dwierichs , I think now everything is updated and ready to merge. Please let me know if there are further questions or requests.

**Review by dwierichs — APPROVED** [review]
@zazabap Thanks for the stamina with this contribution! :tada: 
I will request a second review internally and then we can go ahead and merge :)

File: pennylane/drawer/mpldrawer.py [hunk 11]
```diff
@@ -293,12 +293,30 @@ def __init__(self, n_layers, n_wires, c_wires=0, wire_options=None, figsize=None
...
+                line_options = {**global_options, **specific_options}
+
+                # Create Line2D with the combined options
+                line = plt.Line2D(
+                    (-1, self.n_layers),
+                    (wire, wire),
+                    zorder=1,
+                    **line_options,
+                )
+                self._wire_lines.append(line)
```
Comment (astralcai) [inline pennylane/drawer/mpldrawer.py:318]:
```suggestion

        # Separate global options from per wire options
        global_options = {k: v for k, v in wire_options.items() if not isinstance(v, dict)}
        wire_specific_options = {k: v for k, v in wire_options.items() if isinstance(v, dict)}

        # Adding wire lines with individual styles based on wire_options
        self._wire_lines = []
        for wire in range(self.n_wires):
            specific_options = wire_specific_options.get(wire, {})
            line_options = {**global_options, **specific_options}

            # Create Line2D with the combined options
            line = plt.Line2D(
                (-1, self.n_layers),
                (wire, wire),
                zorder=1,
                **line_options,
            )
            self._wire_lines.append(line)
```

File: pennylane/drawer/draw.py [hunk 12]
```diff
@@ -568,6 +571,26 @@ def circuit2(x, y):
                 :width: 60%
                 :target: javascript:void(0);
 
+
+        Additionally, ``wire_options`` may contain wire labels as keys and dictionaries with
+        keyword-value pairs of matplotlib options as values, which will control the line style for
+        the respective individual wire.
```
Comment (astralcai) [inline pennylane/drawer/draw.py:577]: Maybe rephrase it to
```suggestion
        Additionally, ``wire_options`` may contain sub-dictionaries of matplotlib options assigned
        to separate wire labels, which will control the line style for the respective individual wires.
```

File: pennylane/drawer/tape_mpl.py [hunk 13]
```diff
@@ -450,14 +452,31 @@ def tape_mpl(
...
-        fig, ax = tape_mpl(tape, wire_options={'color':'teal', 'linewidth': 5},
+        fig, ax = qml.drawer.tape_mpl(tape, wire_options={'color':'teal', 'linewidth': 5},
                     label_options={'size': 20})
 
     .. figure:: ../../_static/tape_mpl/wires_labels.png
             :align: center
             :width: 60%
             :target: javascript:void(0);
 
+    Additionally, ``wire_options`` may contain wire labels as keys and dictionaries with
```
Comment (astralcai) [inline pennylane/drawer/tape_mpl.py:463]: Same here.

**Review by astralcai — COMMENTED** [review]
Some minor comments

File: pennylane/drawer/tape_mpl.py (hunk 13 above)
Comment (zazabap) [reply pennylane/drawer/tape_mpl.py]: Thanks a lot for the comment, I have updated the necessary function description part and the mpldrawer.py 👍 . I am not sure why some checks failed this time though 😢 .

Comment (dwierichs) [reply pennylane/drawer/tape_mpl.py]: It is a problem with the CI pipeline, not on you! Let's retrigger CI and see whether it passes.

**Comment by dwierichs** [comment]
Hi @zazabap 
We're experiencing problems with CI across the board. I think this PR is essentially merge-ready, so we'll merge it in as soon as we technically can :) 
Thank you for this nice contribution! :tada:

**Review by astralcai — APPROVED** [review]
Thank you for your contribution!
